source/tests/system/nixnetsocket_driver_api_tests.cpp

```diff
@@ -495,19 +495,149 @@ TEST_F(NiXnetSocketLoopbackTests, OpenedStackThenCloseOriginal_GetStackInfoFromO
+  EXPECT_SUCCESS(rx_stack);
+  EXPECT_SUCCESS(wait);
+  EXPECT_SUCCESS(rx_socket);
+  EXPECT_SUCCESS(bind);
+  EXPECT_SUCCESS(get_name);
+  EXPECT_SUCCESS(listen);
+  EXPECT_SUCCESS(tx_socket);
+
+  const auto connect = client::connect(stub(), tx_socket.socket(), get_name.addr());
+  const auto accept = client::accept(stub(), rx_socket.socket());
+  const auto send = client::send(stub(), tx_socket.socket(), MESSAGE, 0);
+  const auto recv = client::recv(stub(), accept.socket(), MESSAGE.size(), 0);
+
+  EXPECT_SUCCESS(connect);
+  EXPECT_SUCCESS_WITH_STATUS(send, MESSAGE.size());
+  EXPECT_SUCCESS_WITH_STATUS(recv, MESSAGE.size());
+  EXPECT_EQ(MESSAGE, recv.mem());
+}
+
+TEST_F(NiXnetSocketLoopbackTests, IPv4TwoWaySession_CloseAllSockets_SocketsAreClosed)
+{
+  const auto MESSAGE = "HelloWorld"s;
+  const auto RX_ADDR = "10.0.0.1"s;
+  const auto rx_stack = client::ip_stack_create(stub(), "", create_simple_config("ENET1", RX_ADDR));
+  const auto tx_stack = client::ip_stack_create(stub(), "", create_simple_config("ENET4", "10.0.0.2"));
   const auto wait = client::ip_stack_wait_for_interface(stub(), rx_stack.stack_ref(), "", 5000);
   const auto rx_socket = socket(stub(), rx_stack.stack_ref());
-  const auto rx_p_to_n = client::inet_p_to_n(stub(), rx_stack.stack_ref(), 2 /* AF INET*/, "127.0.0.1");
+  const auto sock_addr = lookup_ipv4_sock_addr(stub(), rx_stack.stack_ref(), RX_ADDR, 0);
+  const auto bind = client::bind(stub(), rx_socket.socket(), sock_addr);
+  const auto get_name = client::get_sock_name(stub(), rx_socket.socket());
+  const auto listen = client::listen(stub(), rx_socket.socket(), 10);
+  const auto tx_socket = socket(stub(), tx_stack.stack_ref());
+  const auto connect = client::connect(stub(), tx_socket.socket(), get_name.addr());
+  const auto accept = client::accept(stub(), rx_socket.socket());
+  const auto send = client::send(stub(), tx_socket.socket(), MESSAGE, 0);
+  const auto recv = client::recv(stub(), accept.socket(), MESSAGE.size(), 0);
+
+  const auto close_tx = client::close(stub(), tx_socket.socket());
+  const auto close_rx = client::close(stub(), rx_socket.socket());
+  const auto close_accepted = client::close(stub(), accept.socket());
+
+  EXPECT_SUCCESS(close_tx);
+  EXPECT_SUCCESS(close_rx);
+  EXPECT_SUCCESS(close_accepted);
+  EXPECT_XNET_ERROR(GENERIC_NXSOCKET_ERROR, INVALID_SOCKET_ERROR, INVALID_SOCKET_MESSAGE, client::get_sock_name(stub(), tx_socket.socket()));
+  EXPECT_XNET_ERROR(GENERIC_NXSOCKET_ERROR, INVALID_SOCKET_ERROR, INVALID_SOCKET_MESSAGE, client::get_sock_name(stub(), rx_socket.socket()));
+  EXPECT_XNET_ERROR(GENERIC_NXSOCKET_ERROR, INVALID_SOCKET_ERROR, INVALID_SOCKET_MESSAGE, client::get_sock_name(stub(), accept.socket()));
+}
+
+TEST_F(NiXnetSocketLoopbackTests, UdpSocketsOnWiredPhysicalLoopback_SendAndReceiveData_ReceivesExpectedData)
+{
+  const auto MESSAGE = "HelloWorld"s;
+  const auto rx_stack = client::ip_stack_create(stub(), "", create_simple_config("ENET1", "10.0.0.1"));
```

Comment:
Nit: Consider adding `RX_ADDR`/`TX_ADDR` constants to use here and in the lookup calls.